### PR TITLE
Add rule S6442: React's setState hook should only be used in the render function or body of a component

### LINKED ISSUE #3219: Add rule S6442: React's setState hook should only be used in the render function or body of a component
React's setState hook setter function should not be called directly in the body of a component, otherwise it would produce an infinite render loop. This can happen by mistake. Most commonly, the setter function is called from an event handler.

Non compliant code:
```js
import { useState } from "react";

function ShowLanguage() {
    const [language, setLanguage] = useState("fr-FR");

    // Makes an infinite loop
    setLanguage(navigator.language);

    return (
      <section>
        <h1>Your language is {language}!</h1>
        <button onClick={() => setLanguage("fr-FR")}>Je préfère le Français</button>
      </section>
    );
};
```

Compliant code:
```js
import { useState } from "react";

function ShowLanguage() {
    const [language, setLanguage] = useState(navigator.language);

    return (
      <section>
        <h1>Your language is {language}!</h1>
        <button onClick={() => setLanguage("fr-FR")}>Je préfère le Français</button>
      </section>
    );
};
```

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #3219

### CODE REVIEW

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+type SetterCall = estree.CallExpression & {
+  callee: estree.Identifier;
+};
+
+export const rule: Rule.RuleModule = {
+  meta: {
+    messages: {
+      noHookSetterInBody:
+        "React's setState hook should only be used in the render function or body of a component",
```
Comment (saberduck) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:52]: we should change the message to suggest the action which should be taken (to be actionable). Something like "Move the setState hook to the render function or body of a component"

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    /**
+     * Returns the current scope if it corresponds to a React component function or undefined otherwise.
+     */
+    function getReactComponentScope(): Scope | undefined {
+      const scope = context.getScope();
+      const isReact =
+        scope.type === 'function' &&
+        isFunctionNode(scope.block) &&
```
Comment (saberduck) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:110]: this condition seems to be redundant to `scope.type === function`, do we need it?

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        }
+      },
+    };
+  },
+};
+
+/**
+ * Check if the node is an eslint Rule.Node that's to say it has a parent property.
+ */
+function isRuleNode(node: estree.Node): node is Rule.Node {
```
Comment (saberduck) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:188]: I don't think we need a function for this, we can just inline the check for `node.parent`

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts (hunk 2 above)
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:103]: Indeed it is redundant. I will remove it. Thanks.

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts (hunk 1 above)
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts]: Changed

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts (hunk 3 above)
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:177]: The 'isRuleNode' function avoids casting 'node' twice. Without it, the inlined code looks like the below. Do you prefer me to remove it still?
image

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    function getReactComponentScope(): Scope | undefined {
+      const scope = context.getScope();
+      const isReact = isFunctionNode(scope.block) && matches(scope.block, REACT_PATTERN, 1);
+      return isReact ? scope : undefined;
+    }
+
+    /**
+     * Returns the closest enclosing scope of type 'function' or undefined if there isn't any.
+     */
+    function isInsideFunction(scope: Scope | undefined): boolean {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:104]: I think we should stick to only `null` or  only `undefined` (looking at signatures of these functions)

Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:104]: I would rename parameter `scope` to `functionScope`

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6442.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
+{
+  "title": "React\u0027s setState hook should only be used in the render function or body of a component",
+  "type": "BUG",
+  "status": "ready",
+  "remediation": {
+    "func": "Constant\/Issue",
+    "constantCost": "5min"
+  },
+  "tags": [],
```
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6442.json:9]: the `react` tag is missing

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+  const [, forceUpdate] = React.useReducer(x => x + 1, 0);
+  return forceUpdate;
+}
+
+// Function Declarations..............
+
+function ShowLanguageInvalid() {
+  const [language, setLanguage] = useState("fr-FR");
+
+  // Makes an infinite loop
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:15]: remove those comments

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+  setLanguage(navigator.language); // Noncompliant {{Move the setState hook to the render function or body of a component}}
+//^^^^^^^^^^^
+
+  return (
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:19]: I would simplify everything which is possible in unit test cases, we need such a big jsx

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+
+    return (
+        <section>
+            <h1>Your language is {language}!</h1>
+            <button onClick={() => setLanguage("fr-FR")}>Je préfère le Français</button>
+        </section>
+    );
+}
+
+function ShowLanguageValid() {
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:42]: in fact we tested it already in previous cases

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+  );
+}
+
+// Arrow Functions..............
+
+const ShowLanguageInvalidArrow = () => {
+  const [language, setLanguage] = useState("fr-FR");
+
+  // Makes an infinite loop
+  setLanguage(navigator.language); // Noncompliant {{Move the setState hook to the render function or body of a component}}
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:59]: fyi there is no need to assert message and location everytime

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    <section>
+      <h1>Your language is {language}!</h1>
+      <button onClick={() => setLanguage("fr-FR")}>Je préfère le Français</button>
+    </section>
+  );
+};
+
+// MemberInvalid..............
+
+exports.MemberInvalid = () => {
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:111]: isn't it just arrow function?

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+        }
+      },
+    };
+  },
+};
+
+/**
+ * Check if the node is an eslint Rule.Node that's to say it has a parent property.
+ */
+function isRuleNode(node: estree.Node): node is Rule.Node {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:176]: just rename `hasParent` and no need for any comment

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+ * <ul>
+ *   <li>Identifier: it uses the name
+ *   <li>FunctionDeclaration: it uses the id.
+ *   <li>VariableDeclarator: it uses the id.
+ *   <li>AssignmentExpression: it uses the left part.
+ *   <li>MemberExpression: it uses the property part.
+ * </ul>
+ * If the node is not supported calls itself recursively on the parent node up to max level.
+ */
+function matches(node: estree.Node | null, pattern: RegExp, max = 0): node is estree.Identifier {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:191]: you only call this one with react regex, I would remove this parameter to make the semantic of the function explicit (`isReactComponent`)

Comment (vilchik-elena) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:191]: that's a weird return type... it's not true only when node is Identifier

**Review by vilchik-elena — DISMISSED** [review]
Try to keep to minimum the amount of comments, try to find good variable names, so that just reading the code one can understand what's going on

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+  getModuleNameOfImportedIdentifier,
+  getVariableFromName,
+  isFunctionNode,
+  isIdentifier,
+  isMemberExpression,
+} from '../utils';
+import Variable = EsLintScope.Variable;
+import Scope = EsLintScope.Scope;
+
+// Types used in the hook declaration callback signature which reflects the expectations coming from the selector.
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:34]: Ok, I was not too sure the intent for this types would be clear. So I guess I will just remove it then.

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+     * module name. In the example below <code>isReactName(node)</code> will return true.
+     * @example
+     * import { useState } from "react";
+     *
+     * function ShowLanguageInvalid() {
+     *   const [language, setLanguage] = useState('fr-FR'); // 'useState' id is a React name.
+     *   // [...]
+     * }
+     */
+    function isReactName(identifier: estree.Identifier): boolean {
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:70]: Ok, I will remove it then.

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+     * Returns the current scope if it corresponds to a React component function or undefined otherwise.
+     */
+    function getReactComponentScope(): Scope | undefined {
+      const scope = context.getScope();
+      const isReact = isFunctionNode(scope.block) && matches(scope.block, REACT_PATTERN, 1);
+      return isReact ? scope : undefined;
+    }
+
+    /**
+     * Returns the closest enclosing scope of type 'function' or undefined if there isn't any.
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:102]: As suggested above, I will simply remove the comments.

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+};
+
+type SetterCall = estree.CallExpression & {
+  callee: estree.Identifier;
+};
+
+export const rule: Rule.RuleModule = {
+  meta: {
+    messages: {
+      noHookSetterInBody: 'Move the setState hook to the render function or body of a component',
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:50]: Updating to this until there's a better version: "Move the state setter call to a callback"

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://sonarsource.github.io/rspec/#/rspec/S6442/javascript
+
+import { Rule, Scope as EsLintScope } from 'eslint';
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:22]: Renaming to ESLintScope...

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6442.json (hunk 5 above)
Comment (francoismora) [reply javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S6442.json:9]: Added. What's the goal of this tag?

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js (hunk 6 above)
Comment (francoismora) [reply eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:15]: Removed

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js (hunk 8 above)
Comment (francoismora) [reply eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:42]: Ok, removing

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js (hunk 9 above)
Comment (francoismora) [reply eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:59]: Keeping only the first one.

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts (hunk 11 above)
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:176]: Renamed. I would have imagined that a type predicate should be named isXxx() in general.

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts (hunk 12 above)
Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:191]: Inlined.

Comment (francoismora) [reply eslint-bridge/src/rules/no-hook-setter-in-body.ts:191]: Indeed, that's an error. Thanks.

File: eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js (hunk 10 above)
Comment (francoismora) [reply eslint-bridge/tests/rules/comment-based/no-hook-setter-in-body.js:74]: The implementation is different if this is an 'AssignmentExpression' or a 'VariableDeclarator' as we need to check the name the arrow function is assigned to. So the test aim at covering both cases.

**Review by gabriel-vivas-sonarsource — APPROVED** [review]
The rule description LGTM :+1:

File: eslint-bridge/src/rules/no-hook-setter-in-body.ts [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+
+const REACT_MODULE = 'react';
+const REACT_ROOT = 'React';
+const REACT_PATTERN = /^[^a-z]/;
+const HOOK_FUNCTION = 'useState';
+
+export const rule: Rule.RuleModule = {
+  meta: {
+    messages: {
+      noHookSetterInBody: 'Move the state setter call to a callback',
```
Comment (gabriel-vivas-sonarsource) [inline eslint-bridge/src/rules/no-hook-setter-in-body.ts:53]: Maybe:
```diff
- noHookSetterInBody: 'Move the state setter call to a callback',
+ noHookSetterInBody: 'Remove this state setter call, perhaps move it to an event handler or JSX attribute',
```

**Review by gabriel-vivas-sonarsource — APPROVED** [review]
Reviewed the description and messages, some small tweaks, otherwise LGTM 👍

**Pull request merged**